modules/sov-ibc/src/rpc/helpers.rs

```diff
@@ -40,3 +43,69 @@ pub fn to_jsonrpsee_error(err: impl ToString) -> ErrorObjectOwned {
         None::<String>,
     )
 }
+
+/// Trait for proof agnostic storage value retrieval.
+pub trait StorageValue<V> {
+    type Output;
+    fn value_at_key<K, C, S>(
+        key: &K,
+        store: &StateMap<K, V, C>,
+        working_set: &mut WorkingSet<S>,
+    ) -> Self::Output
+    where
+        S: Spec,
+        C: StateCodec,
+        <C as StateCodec>::ValueCodec: StateItemCodec<V>,
+        <C as StateCodec>::KeyCodec: StateItemCodec<K>;
+}
+
+/// Implementation of [`StorageValue`] for values without proofs.
+pub struct WithoutProof<V>(PhantomData<V>);
+
+impl<V> StorageValue<V> for WithoutProof<V> {
+    type Output = Option<V>;
+
+    fn value_at_key<K, C, S>(
+        key: &K,
+        store: &StateMap<K, V, C>,
+        working_set: &mut WorkingSet<S>,
+    ) -> Self::Output
+    where
+        S: Spec,
+        C: StateCodec,
+        <C as StateCodec>::ValueCodec: StateItemCodec<V>,
+        <C as StateCodec>::KeyCodec: StateItemCodec<K>,
+    {
+        store.get(key, working_set)
+    }
+}
+
+/// Implementation of [`StorageValue`] for values with proofs.
+pub struct WithProof<V>(PhantomData<V>);
+
+impl<V> StorageValue<V> for WithProof<V> {
+    type Output = (Option<V>, Vec<u8>);
+
+    fn value_at_key<K, C, S>(
+        key: &K,
+        store: &StateMap<K, V, C>,
+        working_set: &mut WorkingSet<S>,
+    ) -> Self::Output
+    where
+        S: Spec,
+        C: StateCodec,
+        <C as StateCodec>::ValueCodec: StateItemCodec<V>,
+        <C as StateCodec>::KeyCodec: StateItemCodec<K>,
+    {
+        let result = store.get_with_proof(key, working_set);
+
+        (
+            result
+                .value
+                // if panics, somehow an invalid value was stored in the map
+                .map(|bytes| store.codec().value_codec().decode_unwrap(bytes.value())),
+            // if panics, somehow an invalid proof is returned
+            result.proof.try_to_vec().expect("no error"),
```

Comment:
Panicking might halt the RPC server. Let's change the return type to `RpcResult<V>` so that can propagate this potential error and include it in the response of a query for a better user experience.

Reply:
a2c1956